core/base/vector_cache.cpp

```diff
@@ -0,0 +1,56 @@
+// SPDX-FileCopyrightText: 2017 - 2024 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "ginkgo/core/base/vector_cache.hpp"
+
+#include <ginkgo/core/base/mpi.hpp>
+#include <ginkgo/core/distributed/vector.hpp>
+
+namespace gko {
+namespace detail {
+
+
+template <typename ValueType>
+void VectorCache<ValueType>::init(std::shared_ptr<const Executor> exec,
+                                  gko::experimental::mpi::communicator comm,
+                                  dim<2> global_size, dim<2> local_size) const
+{
+    if (!vec || vec->get_size() != global_size || vec->get_executor() != exec) {
+        vec = experimental::distributed::Vector<ValueType>::create(
+            exec, comm, global_size, local_size);
+    } else if (vec->get_local_vector()->get_size() != local_size) {
+        // handle locally to eliminate the mpi call
+        std::cout << "local diff" << comm.rank() << std::endl;
+        vec->local_ =
+            std::move(gko::matrix::Dense<ValueType>(exec, local_size));
```

Comment:
nit: use the create function instead ?

Reply:
We hold the object not the pointer.
However, I can create the unique pointer and release the object out actually, right?

Reply:
No, It is a only a nit from my side. Using the constructor should also be okay.